File: solutions/scala/149-counting-bits.scala
/**
 * Problem 149: Counting Bits (LeetCode 338)
 * Difficulty: Easy
 * Language: Scala
 */
def countBits(n: Int): Array[Int] =
  (0 to n).foldLeft(Array.empty[Int]) { (dp, i) =>
    dp :+ (if (i == 0) 0 else dp(i >> 1) + (i & 1))
  }

File: solutions/haskell/149-counting-bits.hs
{-
  Problem 149: Counting Bits (LeetCode 338)
  Difficulty: Easy
  Language: Haskell
-}
countBits :: Int -> [Int]
countBits n = foldl (\dp i ->
  dp ++ [if i == 0 then 0 else (dp !! (i `div` 2)) + (i `mod` 2)]
  ) [] [0..n]
